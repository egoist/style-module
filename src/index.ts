let _id: number
let sheet: StyleSheet | null

function reset() {
  _id = 0
  sheet = document.head.appendChild(document.createElement('style')).sheet
}

reset()

function kebabcase(str: string) {
  return str.replace(/[A-Z]/g, '-$&').toLowerCase()
}

function insert(rule: string) {
  if (sheet instanceof CSSStyleSheet) {
    sheet.insertRule(rule, sheet.cssRules.length)
  }
}

interface RuleObj {
  [k: string]: string | RuleObj
}

function css(obj: RuleObj) {
  const id = `sm_${_id++}`
  const { rules, composes } = parse(obj, `.${id}`)
  for (const rule of rules) {
    insert(rule)
  }
  if (composes) {
    return `${composes} ${id}`
  }
  return id
}

function wrap(stringToWrap: string, wrapper: string) {
  return wrapper + '{' + stringToWrap + '}'
}

function parse(obj: RuleObj, className: string, isInsideObj = false) {
  const rules = ['']
  let composes

  for (let prop of Object.keys(obj)) {
    let value = obj[prop]

    if (prop === 'composes' && typeof value === 'string') {
      composes = value
      continue
    }

    prop = kebabcase(prop)
    if (typeof value === 'object') {
      if (/^(:|>|\.|\*)/.test(prop)) {
        prop = className + prop
      }
      // replace & in "&:hover", "p>&"
      prop = prop.replace(/&/g, className)
      const res = parse(value, className, !/^@/.test(prop))
      rules.push(wrap(res.rules.join(''), prop))
    } else {
      rules[0] += `${prop}:${value};`
    }
  }
  if (!isInsideObj) {
    rules[0] = wrap(rules[0], className)
  }
  return { rules, composes }
}

interface StyleModuleInput {
  [k: string]: RuleObj
}

interface StyleModuleOutput {
  [k: string]: string
}

function styleModule(
  obj: StyleModuleInput | (() => StyleModuleInput)
): StyleModuleOutput {
  const decl = typeof obj === 'function' ? obj() : obj
  return Object.keys(decl).reduce((res: StyleModuleOutput, className) => {
    res[className] = css(decl[className])
    return res
  }, {})
}

export { css, styleModule, sheet, reset }
