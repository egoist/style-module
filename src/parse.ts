interface RuleObj {
  [k: string]: string | RuleObj
}

function kebabcase(str: string) {
  return str.replace(/[A-Z]/g, '-$&').toLowerCase()
}

function wrap(stringToWrap: string, wrapper: string) {
  return wrapper + '{' + stringToWrap + '}'
}

function parse(obj: RuleObj, className: string, isInsideObj = false) {
  const rules = ['']

  for (let prop of Object.keys(obj)) {
    let value = obj[prop]

    if (prop === 'composes') {
      continue
    }

    prop = kebabcase(prop)
    if (typeof value === 'object') {
      if (/^(:|>|\.|\*)/.test(prop)) {
        prop = className + prop
      }
      // replace & in "&:hover", "p>&"
      prop = prop.replace(/&/g, className)
      const res = parse(value, className, !/^@/.test(prop))
      rules.push(wrap(res.join(''), prop))
    } else {
      rules[0] += `${prop}:${value};`
    }
  }
  if (!isInsideObj) {
    rules[0] = wrap(rules[0], className)
  }
  return rules
}

export {
  RuleObj,
  parse,
  wrap
}
