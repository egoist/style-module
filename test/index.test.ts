import { css, sheet, styleModule, reset } from '../src'

const getStyle = () => {
  let res = ''
  if (sheet instanceof CSSStyleSheet) {
    for (let i = 0; i < sheet.cssRules.length; i++) {
      res += sheet.cssRules[i].cssText
    }
  }
  return res
}

beforeEach(() => {
  reset()
})

describe('css', () => {
  test('simple', () => {
    expect(
      css({
        color: 'red',
        fontSize: '16px',
        ':hover': {
          color: 'blue'
        },
        '& .foo': {
          color: 'yellow'
        }
      })
    ).toBe('sm_0')
    expect(getStyle()).toMatchSnapshot()
  })

  test('composes', () => {
    const black = css({
      color: 'black'
    })
    const className = css({
      composes: black,
      color: 'red'
    })
    expect(className).toBe('sm_0 sm_1')
    expect(getStyle()).toMatchSnapshot()
  })

  // test('nesting', () => {
  //   const className = css({
  //     '& .button': {
  //       color: 'pink',
  //       ':hover': {
  //         color: 'red'
  //       }
  //     }
  //   })
  //   expect(className).toBe('sm_0')
  //   expect(getStyle()).toMatchSnapshot()
  // })
})

test('styleModule', () => {
  const styles = styleModule({
    button: {
      color: 'red',
      fontSize: '16px',
      ':hover': {
        color: 'blue'
      },
      '& .foo': {
        color: 'yellow'
      }
    }
  })
  expect(styles.button).toBe('sm_0')
  expect(getStyle()).toMatchSnapshot()
})
