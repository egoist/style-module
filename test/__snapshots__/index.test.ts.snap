// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css composes 1`] = `".sm_0 {color: black;}.sm_1 {color: red;}"`;

exports[`css simple 1`] = `".sm_0 {color: red; font-size: 16px;}.sm_0:hover {color: blue;}.sm_0 .foo {color: yellow;}"`;

exports[`styleModule 1`] = `".sm_0 {color: red; font-size: 16px;}.sm_0:hover {color: blue;}.sm_0 .foo {color: yellow;}"`;
